src/coincident/search/opentopo_api.py

```diff
@@ -0,0 +1,179 @@
+https://portal.opentopography.org/apidocs/
+There is also a GitHub repo containing flight extent polygons for NOAA
+https://github.com/OpenTopography/Data_Catalog_Spatial_Boundaries/tree/main/NOAA_Coastal_Lidar
+"""
+
+from __future__ import annotations
+
+from datetime import datetime
+
+import geopandas as gpd
+import pandas as pd
+import requests
+from pandas import Timestamp
+from shapely.geometry import shape
+
+
+def search_ncalm_noaa(
+    aoi: gpd.GeoDataFrame | gpd.GeoSeries,
+    search_start: Timestamp,
+    search_end: Timestamp,
+    dataset: str | None = None,
+) -> gpd.GeoDataFrame:
+    """
+    Perform a search for geospatial data using OpenTopography API.
+    This function dynamically adjusts the API URL based on the dataset.
+
+    Parameters
+    ----------
+    aoi : gpd.GeoDataFrame | gpd.GeoSeries
+        A GeoDataFrame or GeoSeries containing a geometry to restrict the search area.
+    search_start : Timestamp, optional
+        The start datetime for the search, by default None.
+    search_end : Timestamp, optional
+        The end datetime for the search, by default None.
+    dataset : str
+        The dataset type (either "noaa" or "ncalm").
+
+    Returns
+    -------
+    gpd.GeoDataFrame
+        A GeoDataFrame containing the search results.
+
+    Raises
+    ------
+    ValueError
+        If no valid `aoi` is provided or the API request fails.
+    """
+    # If `aoi` is None, set the entire world as the search area
+    if aoi is None:
+        search_poly = gpd.GeoSeries(
+            [
+                shape(
+                    {
+                        "type": "Polygon",
+                        "coordinates": [
+                            [
+                                [-180, -90],
+                                [180, -90],
+                                [180, 90],
+                                [-180, 90],
+                                [-180, -90],
+                            ]
+                        ],
+                    }
+                )
+            ]
+        )
+    else:
+        # convex_hull works better than simplify for more-complex geometry (ie. Louisiana)
+        # https://raw.githubusercontent.com/unitedstates/districts/refs/heads/gh-pages/states/LA/shape.geojson
+        search_poly = aoi.to_crs(4326).union_all()
+        search_poly_chull = search_poly.convex_hull
+        coords = ",".join([f"{x},{y}" for x, y in search_poly_chull.exterior.coords])
+
+    # alter the API URL based on the dataset
+    if dataset == "noaa":
+        url = f"https://portal.opentopography.org/API/otCatalog?productFormat=PointCloud&polygon={coords}&detail=true&outputFormat=json&include_federated=true"
+    elif dataset == "ncalm":
+        url = f"https://portal.opentopography.org/API/otCatalog?productFormat=PointCloud&polygon={coords}&detail=true&outputFormat=json&include_federated=false"
```

Comment:
looks like these URLs are identical? Note that you can pass all the 'parameters' after the API root url as a python dictionary - see https://requests.readthedocs.io/en/latest/user/quickstart/#passing-parameters-in-urls

Reply:
Ah great. The only difference between the URLs is the last key `include_federated` being true or false. True includes NOAA and 3dep. I suppose we could just have `include_federated=true` and have a singular base link since I have another if/else statement a few lines later as seen below which makes the two separate URLs a bit redundant.
[here](https://github.com/uw-cryo/coincident/pull/37/files/baafc3f04bf8bb087ef177d396875e13e48f4171#diff-75c19843f61d524d9257b8cad60a33ae5baff7eee6a9fa7b89a0ab5971db9357R92-R102)
```
    if dataset == "noaa":
        catalog = [
            ds
            for ds in catalog["Datasets"]
            if "NOAA" in ds["Dataset"]["identifier"]["propertyID"]
            and ds["Dataset"]["temporalCoverage"] != ""
        ]
    else:
        catalog = [
            ds for ds in catalog["Datasets"] if ds["Dataset"]["temporalCoverage"] != ""
        ]
```

Reply:
https://github.com/uw-cryo/coincident/commit/9c5b1ea06383deafef031ce7c33f723d32f8f2bb